From 7df69c0f79af61b0aa402e688672e38434467aa6 Mon Sep 17 00:00:00 2001
From: David Drysdale <drysdale@google.com>
Date: Wed, 8 Jan 2014 16:14:07 +0000
Subject: [PATCH] Don't override explicitly specified search domains

Only set search domains from /etc/resolv.conf if there isn't a value
already present in the channel.
---
 ares_init.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/ares_init.c b/ares_init.c
index d23c1b0ca33885f91c671beda3f57868e34a4a31..db1f69784d9bc2e3109d9946550e617a4a4bcc36 100644
--- a/ares_init.c
+++ b/ares_init.c
@@ -1153,20 +1153,24 @@ static int init_by_resolv_conf(ares_channel channel)
     FILE *fp;
     size_t linesize;
     int error;
+    int update_domains;
 
     /* Don't read resolv.conf and friends if we don't have to */
     if (ARES_CONFIG_CHECK(channel))
         return ARES_SUCCESS;
 
+    /* Only update search domains if they're not already specified */
+    update_domains = (channel->ndomains == -1);
+
     fp = fopen(PATH_RESOLV_CONF, "r");
     if (fp) {
       while ((status = ares__read_line(fp, &line, &linesize)) == ARES_SUCCESS)
       {
-        if ((p = try_config(line, "domain", ';')))
+        if ((p = try_config(line, "domain", ';')) && update_domains)
           status = config_domain(channel, p);
         else if ((p = try_config(line, "lookup", ';')) && !channel->lookups)
           status = config_lookup(channel, p, "bind", "file");
-        else if ((p = try_config(line, "search", ';')))
+        else if ((p = try_config(line, "search", ';')) && update_domains)
           status = set_search(channel, p);
         else if ((p = try_config(line, "nameserver", ';')) &&
                  channel->nservers == -1)
-- 
1.9.3

